query-engine/core/src/schema_builder/input_types/objects/create_objects.rs

```diff
@@ -0,0 +1,286 @@
+                    Some(input_field)
+                } else {
+                    Some(input_field.optional())
+                }
+            }
+        })
+        .collect()
+}
+
+fn field_should_be_kept_for_checked_create_input_type(field: &ScalarFieldRef) -> bool {
+    !field.is_auto_generated_int_id
+}
+
+/// Builds the create input type (<x>UncheckedCreateInput / <x>UncheckedCreateWithout<y>Input)
+/// Also valid for nested inputs. A nested input is constructed if the `parent_field` is provided.
+/// "Unchecked" input refers to allowing to write _all_ scalars on a model directly, which can
+/// lead to unintended data integrity violations if used incorrectly.
+fn unchecked_create_input_type(
+    ctx: &mut BuilderContext,
+    model: &ModelRef,
+    parent_field: Option<&RelationFieldRef>,
+) -> InputObjectTypeWeakRef {
+    // We allow creation from both sides of the relation - which would lead to an endless loop of input types
+    // if we would allow to create the parent from a child create that is already a nested create.
+    // To solve it, we remove the parent relation from the input ("Without<Parent>").
+    let name = match parent_field.map(|pf| pf.related_field()) {
+        Some(ref f) => format!(
+            "{}UncheckedCreateWithout{}Input",
+            model.name,
+            capitalize(f.name.as_str())
+        ),
+        _ => format!("{}UncheckedCreateInput", model.name),
+    };
+
+    return_cached_input!(ctx, &name);
+
+    let input_object = Arc::new(init_input_object_type(name.clone()));
+    ctx.cache_input_type(name, input_object.clone());
+
+    let linking_fields = if let Some(parent_field) = parent_field {
+        let child_field = parent_field.related_field();
+        if child_field.is_inlined_on_enclosing_model() {
+            child_field.linking_fields().scalar_fields().collect()
+        } else {
+            vec![]
+        }
+    } else {
+        vec![]
+    };
+
+    let scalar_fields: Vec<ScalarFieldRef> = model
+        .fields()
+        .scalar()
+        .into_iter()
+        .filter(|sf| !linking_fields.contains(sf))
+        .collect();
+
+    let mut fields = input_fields::scalar_input_fields(
+        ctx,
+        model.name.clone(),
+        "Create",
+        scalar_fields,
+        |_, f: ScalarFieldRef, default: Option<DefaultValue>| {
+            let typ = map_scalar_input_type(&f);
+            if f.is_required && f.default_value.is_none() && (f.is_created_at() || f.is_updated_at()) {
+                input_field(f.name.clone(), typ, default)
+                    .optional()
+                    .nullable_if(!f.is_required)
+            } else if f.is_required && f.default_value.is_none() {
+                input_field(f.name.clone(), typ, default)
+            } else {
+                input_field(f.name.clone(), typ, default)
+                    .optional()
+                    .nullable_if(!f.is_required)
+            }
```

Comment:
This block looks like it could be simplified quite a bit. Maybe something like:
```
input_field(f.name.clone(), typ, default)
                    .optional_if(!f.is_required || f.is_created_at() || f.is_updated_at())
                    .nullable_if(!f.is_required)
```

Reply:
Code is from before the new utilities, will change that.